UI/GamepadEmu.cpp

```diff
@@ -318,53 +318,53 @@ void PSPDpad::ProcessTouch(float x, float y, bool down) {
 
 	float dx = (x - bounds_.centerX()) * inv_stick_size;
 	float dy = (y - bounds_.centerY()) * inv_stick_size;
-	float rad = sqrtf(dx*dx + dy*dy);
+	float rad = sqrtf(dx * dx + dy * dy);
 	if (rad < deadzone || fabs(dx) > 0.5f || fabs(dy) > 0.5)
 		down = false;
 
 	int ctrlMask = 0;
-	int lastDown = down_;
-
 	bool fourWay = g_Config.bDisableDpadDiagonals || rad < 0.2f;
 	if (down) {
 		if (fourWay) {
 			int direction = (int)(floorf((atan2f(dy, dx) / (2 * M_PI) * 4) + 0.5f)) & 3;
 			switch (direction) {
-			case 0: ctrlMask |= CTRL_RIGHT; break;
-			case 1: ctrlMask |= CTRL_DOWN; break;
-			case 2: ctrlMask |= CTRL_LEFT; break;
-			case 3: ctrlMask |= CTRL_UP; break;
+			case 0: ctrlMask = CTRL_RIGHT; break;
+			case 1: ctrlMask = CTRL_DOWN; break;
+			case 2: ctrlMask = CTRL_LEFT; break;
+			case 3: ctrlMask = CTRL_UP; break;
 			}
 			// 4 way pad
 		} else {
 			// 8 way pad
 			int direction = (int)(floorf((atan2f(dy, dx) / (2 * M_PI) * 8) + 0.5f)) & 7;
 			switch (direction) {
-			case 0: ctrlMask |= CTRL_RIGHT; break;
-			case 1: ctrlMask |= CTRL_RIGHT | CTRL_DOWN; break;
-			case 2: ctrlMask |= CTRL_DOWN; break;
-			case 3: ctrlMask |= CTRL_DOWN | CTRL_LEFT; break;
-			case 4: ctrlMask |= CTRL_LEFT; break;
-			case 5: ctrlMask |= CTRL_UP | CTRL_LEFT; break;
-			case 6: ctrlMask |= CTRL_UP; break;
-			case 7: ctrlMask |= CTRL_UP | CTRL_RIGHT; break;
+			case 0: ctrlMask = CTRL_RIGHT; break;
+			case 1: ctrlMask = CTRL_RIGHT | CTRL_DOWN; break;
+			case 2: ctrlMask = CTRL_DOWN; break;
+			case 3: ctrlMask = CTRL_DOWN | CTRL_LEFT; break;
+			case 4: ctrlMask = CTRL_LEFT; break;
+			case 5: ctrlMask = CTRL_UP | CTRL_LEFT; break;
+			case 6: ctrlMask = CTRL_UP; break;
+			case 7: ctrlMask = CTRL_UP | CTRL_RIGHT; break;
 			}
 		}
 	}
 
+	int lastDown = down_;
 	down_ = ctrlMask;
-	int pressed = down_ & ~lastDown;
-	int released = (~down_) & lastDown;
-	static const int dir[4] = {CTRL_RIGHT, CTRL_DOWN, CTRL_LEFT, CTRL_UP};
-	for (int i = 0; i < 4; i++) {
-		if (pressed & dir[i]) {
-			if (g_Config.bHapticFeedback) {
-				System_Vibrate(HAPTIC_VIRTUAL_KEY);
+	__CtrlUpdateButtons(ctrlMask, CTRL_LEFT | CTRL_RIGHT | CTRL_UP | CTRL_DOWN);
```

Comment:
I guess this makes the assumption that you would never use two inputs at once that affect the dpad.  Before it was careful to only release things that were changed, but now any touch to the d-pad will FORCE release any other directions.

So if you have a combo key that presses say, up and triangle for some reason - and then you press left, the up button will be force released and discarded.

Is that intentional?  I don't know if any games might motivate that sort of combo button, but it doesn't seem impossible at all.

-[Unknown]

Reply:
Oh, hm, I overlookee that. Might change it back

Reply:
Noticed I forgot to fix this. Fix going into 1.16.1